File: backend/internal/models/invoice.go
package models

import (
	"time"

	"github.com/google/uuid"
)

type InvoiceStatus string

const (
	InvoiceStatusDraft     InvoiceStatus = "draft"
	InvoiceStatusSent      InvoiceStatus = "sent"
	InvoiceStatusPaid      InvoiceStatus = "paid"
	InvoiceStatusOverdue   InvoiceStatus = "overdue"
	InvoiceStatusCancelled InvoiceStatus = "cancelled"
)

type Invoice struct {
	ID            uuid.UUID     `gorm:"type:uuid;primary_key;default:gen_random_uuid()" json:"id"`
	XeroInvoiceID string        `gorm:"uniqueIndex;size:255" json:"xero_invoice_id"` // Xero source of truth
	YachtID       uuid.UUID     `gorm:"type:uuid;not null;index" json:"yacht_id"`
	UserID        uuid.UUID     `gorm:"type:uuid;not null;index" json:"user_id"`
	InvoiceNumber string        `gorm:"uniqueIndex;size:100;not null" json:"invoice_number"`
	Description   string        `gorm:"type:text" json:"description"`
	Amount        float64       `gorm:"type:decimal(10,2);not null" json:"amount"`
	DueDate       time.Time     `gorm:"type:date;index" json:"due_date"`
	Status        InvoiceStatus `gorm:"type:varchar(20);not null;index;default:'draft'" json:"status"`
	IssuedDate    time.Time     `gorm:"type:date" json:"issued_date"`
	PaidDate      *time.Time    `gorm:"type:date" json:"paid_date,omitempty"`
	XeroSyncedAt  *time.Time    `json:"xero_synced_at,omitempty"`
	CreatedAt     time.Time     `json:"created_at"`
	UpdatedAt     time.Time     `json:"updated_at"`

	// Relationships
	Yacht Yacht `gorm:"foreignKey:YachtID;constraint:OnDelete:CASCADE" json:"yacht,omitempty"`
	User  User  `gorm:"foreignKey:UserID;constraint:OnDelete:CASCADE" json:"user,omitempty"`
}

func (Invoice) TableName() string {
	return "invoices"
}

// InvoiceViewModel - Aggregated view for invoice dashboard
type InvoiceViewModel struct {
	Stats            InvoiceStats      `json:"stats"`
	Invoices         []InvoiceInfo     `json:"invoices"`
	RecentActivities []InvoiceActivity `json:"recent_activities"`
}

// InvoiceStats - Summary statistics for invoices
type InvoiceStats struct {
	TotalOutstanding float64 `json:"total_outstanding"`
	PaidCount        int     `json:"paid_count"`
	OverdueCount     int     `json:"overdue_count"`
	PendingCount     int     `json:"pending_count"`
	DraftCount       int     `json:"draft_count"`
}

// InvoiceInfo - Simplified invoice data for list view
type InvoiceInfo struct {
	ID            uuid.UUID `json:"id"`
	InvoiceNumber string    `json:"invoice_number"`
	Description   string    `json:"description"`
	Amount        float64   `json:"amount"`
	DueDate       time.Time `json:"due_date"`
	IssuedDate    time.Time `json:"issued_date"`
	Status        string    `json:"status"`
	IsOverdue     bool      `json:"is_overdue"`
	DaysUntilDue  int       `json:"days_until_due"`
}

// InvoiceActivity - Recent invoice-related activity
type InvoiceActivity struct {
	Icon     string    `json:"icon"`
	Title    string    `json:"title"`
	Subtitle string    `json:"subtitle"`
	Time     time.Time `json:"time"`
	Color    string    `json:"color"`
}

File: iOS/YachtLife/YachtLife/Models/Invoice.swift
import Foundation
import SwiftUI

struct Invoice: Codable, Identifiable {
    let id: UUID
    let yachtId: UUID
    let userId: UUID
    let xeroInvoiceId: String?
    let amount: Double
    let dueDate: Date
    let status: InvoiceStatus
    let description: String
    let createdAt: Date
    let paidAt: Date?

    enum CodingKeys: String, CodingKey {
        case id
        case yachtId = "yacht_id"
        case userId = "user_id"
        case xeroInvoiceId = "xero_invoice_id"
        case amount
        case dueDate = "due_date"
        case status
        case description
        case createdAt = "created_at"
        case paidAt = "paid_at"
    }

    enum InvoiceStatus: String, Codable {
        case draft
        case sent
        case paid
        case overdue
        case cancelled
    }
}

struct Payment: Codable, Identifiable {
    let id: UUID
    let invoiceId: UUID
    let userId: UUID
    let amount: Double
    let paymentMethod: PaymentMethod
    let stripePaymentId: String?
    let xeroPaymentId: String?
    let status: PaymentStatus
    let createdAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case invoiceId = "invoice_id"
        case userId = "user_id"
        case amount
        case paymentMethod = "payment_method"
        case stripePaymentId = "stripe_payment_id"
        case xeroPaymentId = "xero_payment_id"
        case status
        case createdAt = "created_at"
    }

    enum PaymentMethod: String, Codable {
        case stripe
        case applePay = "apple_pay"
        case googlePay = "google_pay"
        case bankTransfer = "bank_transfer"
    }

    enum PaymentStatus: String, Codable {
        case pending
        case completed
        case failed
        case refunded
    }
}

// MARK: - Invoice ViewModel (Dashboard)

struct InvoiceViewModel: Codable {
    let stats: InvoiceStats
    let invoices: [InvoiceInfo]
    let recentActivities: [InvoiceActivity]

    enum CodingKeys: String, CodingKey {
        case stats
        case invoices
        case recentActivities = "recent_activities"
    }
}

struct InvoiceStats: Codable {
    let totalOutstanding: Double
    let paidCount: Int
    let overdueCount: Int
    let pendingCount: Int
    let draftCount: Int

    enum CodingKeys: String, CodingKey {
        case totalOutstanding = "total_outstanding"
        case paidCount = "paid_count"
        case overdueCount = "overdue_count"
        case pendingCount = "pending_count"
        case draftCount = "draft_count"
    }
}

struct InvoiceInfo: Codable, Identifiable {
    let id: UUID
    let invoiceNumber: String
    let description: String
    let amount: Double
    let dueDate: Date
    let issuedDate: Date
    let status: String
    let isOverdue: Bool
    let daysUntilDue: Int

    enum CodingKeys: String, CodingKey {
        case id
        case invoiceNumber = "invoice_number"
        case description
        case amount
        case dueDate = "due_date"
        case issuedDate = "issued_date"
        case status
        case isOverdue = "is_overdue"
        case daysUntilDue = "days_until_due"
    }

    var statusColor: Color {
        switch status.lowercased() {
        case "paid": return .green
        case "sent": return .blue
        case "overdue": return .red
        case "draft": return .gray
        default: return .gray
        }
    }

    var statusIcon: String {
        switch status.lowercased() {
        case "paid": return "checkmark.circle.fill"
        case "sent": return "envelope.fill"
        case "overdue": return "exclamationmark.triangle.fill"
        case "draft": return "doc.text"
        default: return "doc"
        }
    }

    var formattedAmount: String {
        return String(format: "$%.2f", amount)
    }

    var formattedDueDate: String {
        let formatter = DateFormatter()
        formatter.dateFormat = "MMM d, yyyy"
        return formatter.string(from: dueDate)
    }

    var dueDateInfo: String {
        if isOverdue {
            return "Overdue"
        } else if daysUntilDue == 0 {
            return "Due today"
        } else if daysUntilDue == 1 {
            return "Due tomorrow"
        } else if daysUntilDue > 0 {
            return "Due in \(daysUntilDue) days"
        } else {
            return "Overdue by \(abs(daysUntilDue)) days"
        }
    }
}

struct InvoiceActivity: Codable, Identifiable {
    let icon: String
    let title: String
    let subtitle: String
    let time: Date
    let color: String

    var id: String {
        "\(icon)_\(time.timeIntervalSince1970)"
    }

    var colorValue: Color {
        switch color {
        case "blue": return .blue
        case "green": return .green
        case "orange": return .orange
        case "red": return .red
        case "gray": return .gray
        default: return .gray
        }
    }

    var timeAgo: String {
        let formatter = RelativeDateTimeFormatter()
        formatter.unitsStyle = .short
        return formatter.localizedString(for: time, relativeTo: Date())
    }
}
